lib/docker-toolbelt.coffee

```diff
@@ -337,13 +337,17 @@ DockerToolbelt::createDeltaAsync = (src, dest, onProgress) ->
 #          { registry: "registry.resinstaging.io", imageName: "resin/rpi" }
 DockerToolbelt::getRegistryAndName = Promise.method (image) ->
 	# Matches (registry)/(repo)(optional :tag or @digest)
-	# The regex for digest is adapted from https://github.com/docker/distribution/blob/95daa793b83a21656fe6c13e6d5cf1c3999108c7/reference/regexp.go#L44
-	match = image.match(/^(?:([^\/:.]+\.[^\/:]+(?::[0-9]+)?)\/)?([^\/:@]+(?:\/[^\/:@]+)?)(?:(?::(.*))|(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*:[0-9a-f-A-F]{32,})))?$/)
+	# regex adapted from Docker's source code:
+	# https://github.com/docker/distribution/blob/release/2.7/reference/normalize.go#L62
+	# https://github.com/docker/distribution/blob/release/2.7/reference/regexp.go#L44
+	match = image.match /^(?:(localhost|.*?[.:].*?)\/)?(.+?)(?::(.*?))?(?:@(.*?))?$/
```

Comment:
* I've split the previous regex in two: this regex has a simple `(?:@(.*?))?` pattern for the digest part, and the digest is further validated below in a digest-specific regex.
* The new regex is less strict for the registry-domain / repo / image / tag parts, using `.*?` instead of `[^\/:.]+`. I don't think this is very important (I'm happy to change it), but it more closely matches at least the domain parsing in the `splitDockerDomain` function in Docker's source code:
  * https://github.com/docker/distribution/blob/release/2.7/reference/normalize.go#L62
  * `'localhost'` is treated as a special case (ugly if you ask me, but that's what Docker does).
  * The previous regex insisted on a `name.com:123` format for the registry-domain, but this ignores for example the IPv6 notation `[::1]:123`.
* I was debating with myself if it's beneficial to fail image name parsing altogether by throwing an error, as opposed to assume that anything that doesn't match the code's expectation should be deemed to be part of the `imageName` component as opposed to the other components, and let operations fail at a higher level (like, allow calling code to attempt to pull the image and potentially fail with "image not found").
* Most importantly, the new regexes pass for all previously existing spec/test cases, plus new test cases that cover 20+ sample image names including the format reported in the issue/ticket. At least in this sense, the new implementation is preferable to the previous one.